src/libopensc/card-piv.c

```diff
@@ -690,88 +2695,91 @@ static int piv_select_aid(sc_card_t* card, u8* aid, size_t aidlen, u8* response,
 
 static int piv_find_aid(sc_card_t * card)
 {
-	sc_apdu_t apdu;
+	piv_private_data_t * priv = PIV_DATA(card);
 	u8 rbuf[SC_MAX_APDU_BUFFER_SIZE];
 	int r,i;
 	const u8 *tag;
 	size_t taglen;
+	const u8 *nextac;
 	const u8 *pix;
 	size_t pixlen;
+	const u8 *actag;  /* Cipher Suite */
+	size_t actaglen;
+	const u8 *csai; /* Cipher Suite Algorithm Identifier */
+	size_t csailen;
 	size_t resplen = sizeof(rbuf);
+#ifdef ENABLE_PIV_SM
+	int found_csai = 0;
+#endif
 
 	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
 
+
 	/* first  see if the default application will return a template
 	 * that we know about.
 	 */
 
 	r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, rbuf, &resplen);
+	if (r > 0 && priv->aid_der.value && resplen == priv->aid_der.len  && !memcmp(priv->aid_der.value, rbuf, resplen)) {
+		LOG_FUNC_RETURN(card->ctx,SC_SUCCESS);
+		/* no need to parse again, same as last time */
```

Comment:
The use of the Discovery object in place of the AID is done for the Yubico device because they failed to implement SELECT AID correctly.  https://github.com/dengert/OpenSC/blob/PIV-4-extensions/src/libopensc/card-piv.c#L526-L529 ,
https://github.com/dengert/OpenSC/blob/PIV-4-extensions/src/libopensc/card-piv.c#L5466-L5499 and https://github.com/dengert/OpenSC/blob/PIV-4-extensions/src/libopensc/card-piv.c#L5547-L5555

[NIST 800-73-4](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf) "2.2 PIV Card Application AID" defines the AID as 'A0 00 00 03 08 00 00 10 00 01 00' where the last '01 00' is a version. This has not changed from previous versions of 800-73, of which only 2 versions are current supported, 800-73-3 and 800-73-4. So although the application can be selected using 'A0 00 00 03 08 00 00 10 00' they have not changed the version number of the applet, because differences in versions can be detected by the presence or absents of certain flags or objects. 800-73-4 is an extension of 800-73-.3. 

If NIST define a new version, we can adapt. 

I can't find it, but I think it says somewhere only one PIV card application per token.

Reply:
thanks for clearification